File: src/converter/plugins/convertUnionType.ts
import ts from "typescript";
import {createSimplePlugin} from "../plugin.js";
import {TypeScriptService, typeScriptServiceKey} from "./TypeScriptPlugin.js";
import {isPossiblyNullableType} from "./NullableUnionTypePlugin.js";

export const convertUnionType = createSimplePlugin((node, context, render) => {
    if (!ts.isUnionTypeNode(node)) return null

    const typeScriptService = context.lookupService<TypeScriptService>(typeScriptServiceKey)

    const isNullable = isPossiblyNullableType(node, context)

    return `Any${isNullable ? "?" : ""} /* ${typeScriptService?.printNode(node)} */`;
})

File: karakum/src/jsMain/kotlin/io/github/sgrishchenko/karakum/extension/plugins/convertUnionType.kt
package io.github.sgrishchenko.karakum.extension.plugins

import io.github.sgrishchenko.karakum.extension.createPlugin
import typescript.isUnionTypeNode

val convertUnionType = createPlugin plugin@{ node, context, _ ->
    if (!isUnionTypeNode(node)) return@plugin null

    val typeScriptService = context.lookupService<TypeScriptService>(typeScriptServiceKey)

    val isNullable = isPossiblyNullableType(node, context)

    "Any${if (isNullable) "?" else ""} /* ${typeScriptService?.printNode(node)} */"
}
